src/meta/src/barrier/mod.rs

```diff
@@ -666,36 +672,37 @@ where
             ));
             let mut notifiers = notifiers;
             notifiers.iter_mut().for_each(Notifier::notify_to_send);
-            checkpoint_control.inject(command_ctx.clone(), notifiers);
 
-            self.inject_and_send_err(command_ctx, barrier_complete_tx.clone())
+            checkpoint_control.enqueue_command(command_ctx.clone(), notifiers);
+            self.inject_barrier_and_collect(command_ctx, barrier_complete_tx.clone())
                 .await;
         }
     }
 
-    /// Inject barrier and send err.
-    async fn inject_and_send_err(
+    /// Inject a barrier to all CNs and try to collect it
+    async fn inject_barrier_and_collect(
         &self,
         command_context: Arc<CommandContext<S>>,
         barrier_complete_tx: UnboundedSender<(u64, MetaResult<Vec<BarrierCompleteResponse>>)>,
     ) {
-        let result = self
-            .inject_barrier(command_context.clone(), barrier_complete_tx.clone())
-            .await;
-        if let Err(e) = result {
-            barrier_complete_tx
-                .send((command_context.prev_epoch.0, Err(e)))
-                .unwrap();
+        let prev_epoch = command_context.prev_epoch.0;
+        let result = self.inject_barrier(command_context.clone()).await;
+        match result {
+            Ok(node_need_collect) => Self::collect_barrier_async(
+                node_need_collect,
+                self.env.stream_client_pool_ref(),
+                command_context,
+                barrier_complete_tx,
+            ),
```

Comment:
Suggest naming the function `**async** fn collect_barrier` and `spawn` here, which makes the property of "asynchronous" clearer.